### DISCUSSION TITLE
Prettify 2D-lists (and dictionary output)

### DISCUSSION THREAD

**Discussion by DieterPi** [dodona-edu/feedback-and-support#52] [Questions - Vragen] [answered]
When working with dictionaries, the default output in Dodona is quite ugly:

`{4: [[2, 2]], 6: [[3, 3]], 8: [[3, 5]], 10: [[3, 7], [5, 5]], 12: [[5, 7]], 14: [[3, 11], [7, 7]], 16: [[3, 13], [5, 11]], 18: [[5, 13], [7, 11]], 20: [[3, 17], [7, 13]], 22: [[3, 19], [5, 17], [11, 11]], 24: [[5, 19], [7, 17], [11, 13]], 26: [[3, 23], [7, 19], [13, 13]], 28: [[5, 23], [11, 17]], 30: [[7, 23], [11, 19], [13, 17]], 32: [[3, 29], [13, 19]], 34: [[3, 31], [5, 29], [11, 23], [17, 17]], 36: [[5, 31], [7, 29], [13, 23], [17, 19]], 38: [[7, 31], [19, 19]], 40: [[3, 37], [11, 29], [17, 23]], 42: [[5, 37], [11, 31], [13, 29], [19, 23]], 44: [[3, 41], [7, 37], [13, 31]], 46: [[3, 43], [5, 41], [17, 29], [23, 23]], 48: [[5, 43], [7, 41], [11, 37], [17, 31], [19, 29]], 50: [[3, 47], [7, 43], [13, 37], [19, 31]], 52: [[5, 47], [11, 41], [23, 29]], 54: [[7, 47], [11, 43], [13, 41], [17, 37], [23, 31]], 56: [[3, 53], [13, 43], [19, 37]], 58: [[5, 53], [11, 47], [17,... `

While I would prefer:

```
{4: [[2,2]],
 6: [[3,3]],
 8: [[3, 5]], 
 10: [[3, 7], [5, 5]], 
 12: [[5, 7]], 
 14: [[3, 11], [7, 7]], 
 16: [[3, 13], [5, 11]],
 ...
```

The same is true for 2d lists:

Now it shows:

```python
>>> transponeer([[6, 7, 6], [3, 7, 7], [7, 6, 4], [7, 4, 3]])
[[6, 3, 7, 7], [7, 7, 6, 4], [6, 7, 4, 3]]
```

While I would prefer:

```python
>>> transponeer([[6, 7, 6], 
                 [3, 7, 7], 
                 [7, 6, 4], 
                 [7, 4, 3]])
[[6, 3, 7, 7], 
 [7, 7, 6, 4], 
 [6, 7, 4, 3]]
```

I know I can write the custom oracle function where the output can be formatted. But are there already some example oracles ready where this is already implemented? Or is there another way to realize this easily?

**Comment by pdawyndt** [comment]
Pretty printing of (nested) collections is on our roadmap for the TESTed judge: https://github.com/dodona-edu/universal-judge/issues/350. 

Once implemented, pretty printing should ideally come automatically. So should not be a concern of exercise authors. We once had this feature in our Python judge, with a result exactly like you’ve shown. We removed it as its implementation did not blend well with other features whose responsability moved from the judge to Dodona (diffing of expected and actual output, syntax highlighting). If done well, it definitely helps students in finding differences between actual and expected return values.

**Reply by DieterPi** [reply]
Thanks. 👍️
Any idea on the timing of this feature? 
(I want to start the second semester with 2D lists, because if needed I will program a custom oracle)

**Reply by pdawyndt** [reply]
@DieterPi For sure it will not be ready by the second semester. We are working together with a master student on some extensions of TESTed, and this is one of the (optional) extensions on his TODO-list (https://github.com/dodona-edu/universal-judge/issues/559). Not something we will work on for the coming few weeks.

**Comment by DieterPi** [comment]
Small follow-up question:

Now I used a custom oracle to format the output as follows:
![afbeelding](https://github.com/user-attachments/assets/6f6e1459-cb8a-45ec-8789-d9aee4aadcf1)

Can I also format the expression? (in a yaml TESTed plan)

**Reply by pdawyndt** [reply, marked as answer]
In addition to an `expression` or a `statement` you can also add a `description` to a testcase. The `expression` / `statement` is what will be evaluated/executed. The `description` is what will be shown to the student when generating feedback.

Try adding a `descripton` with a multiline string to see if that works. Since the description is not parsed, it should not change, but you will loose the syntax highlighting. You can format the `description` using the Python syntax highlighter with

```yaml
description:
  description: "nullen([\n    [11, -6, 1, 4],\n    [-3, -14, -9, 13],\n    [-7, 13, -8, 7],\n     [6, -15, 7, 4],\n    [7, 0, 15, 0],\n    [-5, 8, 12, 9]\n])"
  format: "python"
```

(I haven't check whether the string with newlines should be enclosed in single or double quotes, as this might make a difference in YAML).
